### PR TITLE
moving docs to mkdocs

### LINKED ISSUE #578: remove use of "validation" replace with "parse"
ref #284 and #360 

Pydantic is no a validation library, it's a parsing library.

It makes no guarantee whatsoever about checking the form of data you give to it; instead it makes a guarantee about the form of data you get out of it.

This sounds like an esoteric difference, but it has real practical consequences, eg.
* if you pass `"3"` (which is not an int) to an `int` field, pydantic will convert it to an int
* if you pass `3.14` (which again is not an int) to an `int` field, pydantic will convert it to an int (thereby "loosing information")

I think this is correct and I'm not interested in changing it, but we should be clear about what pydantic is/does - while I was annoyed by the manor of the question in #284 (sorry, wrong issue, I meant #360) I do understand the motivation for the question.

To fix this we should:
* [ ] rename `validate_model` to `parse_model` (this will require the old version to continue to work with a deprecation warning for 2 versions)
* [ ] rename `__get_validators__` to `__get_parser_functions__` or similar (again will require the old function to continue to work with a deprecation warning)
* [ ] add the above statement prominently to the docs
* [ ] remove use of "validate"/"validation" from the docs

This is a big, backwards incompatible change for no material benefit but I think it's worth it for clarity.

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Change Summary

Move all docs to mkdocs

fixes #578

Still to do:
* [x] add an intro to `models.md`
* [x] add an intro to `types.md`
* [x] fix "TODO: warning about field order"
* [x] remove "Required Fields and mypy" section in mypy
* [x] fix benchmarks (should be regenerated for v1 anyway), needs to output an md table not csv

### CODE REVIEW

**Comment by dmontagu** [comment]
This is gonna be really awesome!

**Comment by samuelcolvin** [comment]
this is ready to review, feedback very welcome.

I see need to fix benchmarks before this can be merged, and there's #862 to do later, but unless there are any major problems this should be merged soon since it's already better than what went before.

**Comment by dmontagu** [comment]
I'm looking at it now -- this is awesome!!

File: docs/usage/models.md [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+in an API.
+
+Untrusted data can be passed to a model, after parsing and validation *pydantic* guarantees that the fields
+of the resultant model instance will conform to the field types defined on the model.
+
+!!! note
+    *pydantic* is primarily a parsing library, **not a validation library**.
+    Validation is a means to and end - building a model which forms to the types and constraints provided.
+
+    In other workds *pydantic* guarentees the types and constraints of the output model not the input data.
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
    In other words *pydantic* guarantees the types and constraints of the output model, not the input data.
```

File: docs/usage/models.md [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+You can think of models as similar to types in strictly typed languages or the requirements of a single endpoint
+in an API.
+
+Untrusted data can be passed to a model, after parsing and validation *pydantic* guarantees that the fields
+of the resultant model instance will conform to the field types defined on the model.
+
+!!! note
+    *pydantic* is primarily a parsing library, **not a validation library**.
+    Validation is a means to and end - building a model which forms to the types and constraints provided.
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
    Validation is a means to an end - building a model which conforms to the types and constraints provided.
```

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+```py
+{!./examples/example1.py!}
+```
+_(This script is complete, it should run "as is")_
+
+What's going on here:
+
+* `id` is of type int; the annotation only declaration tells *pydantic* that this field is required. Strings,
+  bytes or floats will be coerced to ints if possible, otherwise an exception would be raised.
+* `name` is inferred as a string from the default, it is not required as it has a default.
```
Comment (dmontagu) [inline docs/index.md]:
```suggestion
* `name` is inferred as a string from the default; it is not required as it has a default.
```

File: docs/index.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+* `signup_ts` is a datetime field which is not required (``None`` if it's not supplied), pydantic will process
+  either a unix timestamp int (e.g. `1496498400`) or a string representing the date & time.
+* `friends` uses python's typing system, it is required to be a list of integers, as with `id` integer-like objects
+  will be converted to integers.
+
+If validation fails pydantic with raise an error with a breakdown of what was wrong:
```
Comment (dmontagu) [inline docs/index.md]:
```suggestion
If validation fails pydantic will raise an error with a breakdown of what was wrong:
```

File: docs/index.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+```
+outputs:
+```json
+{!./examples/example2_output.json!}
+```
+
+## Rationale
+
+
+So *pydantic* uses some cool new language feature, but why should I actually go and use it?
```
Comment (dmontagu) [inline docs/index.md]:
```suggestion
So *pydantic* uses some cool new language features, but why should I actually go and use it?
```

File: docs/index.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+**no brainfuck**
+: no new schema definition micro-language to learn. If you know python (and perhaps skim read the
```
Comment (dmontagu) [inline docs/index.md]:
```suggestion
: no new schema definition micro-language to learn. If you know python (and perhaps skim the
```

File: docs/usage/dataclasses.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+Dataclasses work in python 3.6 using the [dataclasses backport package](https://github.com/ericvsmith/dataclasses).
+
+```py
+{!./examples/ex_dataclasses.py!}
+```
+_(This script is complete, it should run "as is")_
+
+!!! note
+    Keep in mind that `pydantic.dataclasses.dataclass` is a drop in replacement for `dataclasses.dataclass`
+    with validation, not a repacement for `pydantic.BaseModel`. There are cases where subclassing
```
Comment (dmontagu) [inline docs/usage/dataclasses.md]:
```suggestion
    with validation, not a replacement for `pydantic.BaseModel`. There are cases where subclassing
```

File: docs/usage/dataclasses.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    `pydantic.BaseModel` is the better choice. 
+    
+    For more information and disucssion see
```
Comment (dmontagu) [inline docs/usage/dataclasses.md]:
```suggestion
    For more information and discussion see
```

File: docs/usage/models.md [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+
+    In other workds *pydantic* guarentees the types and constraints of the output model not the input data.
+
+    This might sound like an esoteric distinction, but it is not - you should read about 
+    [Data Conversion](#data-conversion) if you're unsure what this means or how it might effect your usage.
+
+## Model basic usage
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
## Basic model usage
```

File: docs/usage/models.md [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+
+```py
+from pydantic import BaseModel
+
+class User(BaseModel):
+    id: int
+    name = 'Jane Doe'
+```
+`User` here is a model with two fields `id` which is an integer and is required, 
+and `name` which is a string and is not required (it has a a default value). The type of `name` is inferred from the
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
and `name` which is a string and is not required (it has a default value). The type of `name` is inferred from the
```

File: docs/usage/models.md [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+the fields which were supplied when user was initialised:
+```py
+assert user.dict() == dict(user) == {'id': 123, 'name': 'Jane Doe'}
+```
+either `.dict()` or `dict(user)` will provide a dict of fields, but `.dict()` can take numerous other arguments.
+```py
+user.id = 321
+assert user.id == 321
+```
+This model is mutable so field can be changed.
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
This model is mutable so field values can be changed.
```

File: docs/usage/models.md [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+
+### Model properties
+
+The example above only shows the tip of the iceberg of what models can do,
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
The example above only shows the tip of the iceberg of what models can do. 
```

File: docs/usage/models.md [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+```py
+user.id = 321
+assert user.id == 321
+```
+This model is mutable so field can be changed.
+
+### Model properties
+
+The example above only shows the tip of the iceberg of what models can do, 
+models contains the following methods and attributes:
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
Models contains the following methods and attributes:
```

File: docs/usage/models.md [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+!!! warning
+    Immutability in python is never strict. If developers are determined/stupid they can always
+    modify a so-called "immutable" object.
+
+```py
+{!./examples/mutation.py!}
+```
+
+Trying to change `a` caused an error and it remains unchanged, however the dict `b` is mutable and the
+immutability of `foobar` doesn't stop being changed.
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
immutability of `foobar` doesn't stop `b` from being changed.
```

File: docs/usage/models.md [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+[Abstract Base Classes](https://docs.python.org/3/library/abc.html) (ABCs).
+
+```py
+{!./examples/ex_abc.py!}
+```
+_(This script is complete, it should run "as is")_
+
+## Field Ordering
+
+Field order is important in models since for the following reason:
```
Comment (dmontagu) [inline docs/usage/models.md]:
```suggestion
Field order is important in models for the following reason:
```

File: docs/usage/types.md [hunk 16]
```diff
@@ -0,0 +1,648 @@
+Where possible *pydantic* uses [standard library types](#standard-library-types) to define fields thus smoothing 
+the learning curve. For some types however no standard library type exists, 
+here *pydantic* implements [many commonly used types](#pydantic-types). If no existing type exists for your 
+purpose you can also implement your [own types](#custom-data-types) with custom properties and validation.
+
+## Standard Library Types
+
+*pydantic* supports many common types from the python standard library, you need stricter processing see
```
Comment (dmontagu) [inline docs/usage/types.md]:
```suggestion
*pydantic* supports many common types from the python standard library. If you need stricter processing see 
```

File: docs/usage/types.md [hunk 17]
```diff
@@ -0,0 +1,648 @@
+Where possible *pydantic* uses [standard library types](#standard-library-types) to define fields thus smoothing 
+the learning curve. For some types however no standard library type exists, 
+here *pydantic* implements [many commonly used types](#pydantic-types). If no existing type exists for your 
+purpose you can also implement your [own types](#custom-data-types) with custom properties and validation.
+
+## Standard Library Types
+
+*pydantic* supports many common types from the python standard library, you need stricter processing see 
+[Strict Types](#strict-types), if you need to constrain the values allowed (eg. require a positive int) see
```
Comment (dmontagu) [inline docs/usage/types.md]:
```suggestion
[Strict Types](#strict-types); if you need to constrain the values allowed (eg. require a positive int) see
```

File: docs/usage/types.md [hunk 18]
```diff
@@ -0,0 +1,648 @@
...
+- `StrictInt` (and the `strict` option of `ConstrainedInt`) will not accept `bool` types,
+    even though `bool` is a subclass of `int` in Python. Other subclasses will work.
+- `StrictFloat` (and the `strict` option of `ConstrainedFloat`) will not accept `int`.
+
+```py
+{!./examples/strict_types.py!}
+```
+_(This script is complete, it should run "as is")_
+
+## Custom Data Types
```
Comment (dmontagu) [inline docs/usage/types.md]: For what it's worth, I personally think this section, short as it is, merits its own top level page, separate from all the other types.

(It may also be worth expanding that page..)

This is such a useful feature and I think it doesn't get the attention it deserves -- it's a shame for it to be buried way at the bottom of the long list of types pydantic provides.

File: docs/usage/model_config.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+**`validate_assignment`**
+: whether to perform validation on assignment to attributes or not (default: `False`)
+
+**`allow_population_by_alias`**
+: whether or not an aliased field may be populated by its name as given by the model
+  attribute, rather than strictly the alias; please be sure to read the warning below before enabling this (default:
+  `False`)
+
+!!! warning
+    Think twice before enabling `allow_population_by_alias`! Enabling it could cause previously correct code to become
```
Comment (dmontagu) [inline docs/usage/model_config.md]: Any chance we could weaken this warning a little bi? Now seems like a good time.

File: docs/usage/settings.md [hunk 20]
```diff
@@ -0,0 +1,47 @@
+One of pydantic's most useful applications is to define default settings, allow them to be overridden by
```
Comment (dmontagu) [inline docs/usage/settings.md]:
```suggestion
One of pydantic's most useful applications is to define default settings, and allow them to be overridden by
```

File: README.md [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+class User(BaseModel):
+    id: int
+    name = 'John Doe'
+    signup_ts: datetime = None
+    friends: List[int] = []
+
+external_data = {'id': '123', 'signup_ts': '2017-06-01 12:22', 'friends': [1, '2', b'3']}
+user = User(**external_data)
+print(user)
+# > User id=123 name='John Doe' signup_ts=datetime.datetime(2017, 6, 1, 12, 22) friends=[1, 2, 3]
```
Comment (samuelcolvin) [inline README.md]:
```suggestion
#> User id=123 name='John Doe' signup_ts=datetime.datetime(2017, 6, 1, 12, 22) friends=[1, 2, 3]
```

File: README.md [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+print(user.id)
+# > 123
```
Comment (samuelcolvin) [inline README.md]:
```suggestion
#> 123
```

File: README.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+```
+
+## Contributing
+
+For guidance on setting up a development environment and how to make a
+contribution to *pydantic*, see the
```
Comment (samuelcolvin) [inline README.md]:
```suggestion
contribution to *pydantic*, see 
```

File: changes/make_history.py [hunk 24]
```diff
@@ -6,10 +6,12 @@
 from pathlib import Path
 
 THIS_DIR = Path(__file__).parent
-name_regex = re.compile(r'(\d+)-(.*?)\.rst')
+name_regex = re.compile(r'(\d+)-(.*?)\.md')
 bullet_list = []
 
-for p in THIS_DIR.glob('*.rst'):
+for p in THIS_DIR.glob('*.md'):
```
Comment (samuelcolvin) [inline changes/make_history.py]:
```suggestion
for p in THIS_DIR.glob('*-*.md'):
```
and below

File: changes/make_history.py [hunk 25]
```diff
@@ -29,20 +31,20 @@
 chunk_title = f'v{version.VERSION} ({date.today():%Y-%m-%d})'
 bullets = '\n'.join(c for *_, c in sorted(bullet_list))
 new_chunk = f"""
-{chunk_title}
-{'.' * len(chunk_title)}
+## {chunk_title}
```
Comment (samuelcolvin) [inline changes/make_history.py]: as one line.

File: docs/build/main.py [hunk 26]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python3
+import re
+import sys
+from pathlib import Path
+
+THIS_DIR = Path(__file__).parent
+
```
Comment (samuelcolvin) [inline docs/build/main.py]:
```suggestion
```

File: docs/contributing.md [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+label on github.
+
+To make contributing as easy and fast as possible, you'll want to run tests and linting locally. Luckily since
+*pydantic* has few dependencies, doesn't require compiling and tests don't need access to databases etc., setting
+up and running tests should be very simple.
+
+You'll need to have **python 3.6** or **3.7**, **virtualenv**, **git**, and **make** installed.
+
+```bash
+{!./examples/contributing.sh!}
```
Comment (samuelcolvin) [inline docs/contributing.md]: may as well copy content to here.

File: docs/usage/model_config.md (hunk 19 above)
Comment (samuelcolvin) [reply docs/usage/model_config.md]: #585 yes but I'll do it as a separate PR, also need to rename it to `allow_population_by_field_name`.

**Comment by samuelcolvin** [comment]
@dmontagu thank you very much for your review. I know this stuff isn't much fun, but I'm very dyslexic and corrects like this make the docs a lots easier to use and looks a lot more professional.

File: docs/usage/types.md (hunk 18 above)
Comment (samuelcolvin) [reply docs/usage/types.md]: okay, personally I think it's right here but we should link to it more and expand it to contain more details and a better example. Certainly the current example is awful.

File: docs/usage/model_config.md (hunk 19 above)
Comment (samuelcolvin) [reply docs/usage/model_config.md]: #875

**Comment by haizaar** [comment]
Guys, can you please elaborate why did you decided to switch from rst/sphinx to mkdocs? Any particular advantages in favor the latter?

**Comment by dmontagu** [comment]
I think the new (hierarchical) organization is a big improvement; it's not just a massive wall of text. I'm not sure that by itself necessitated the move to mkdocs though.

------

I can't speak for what motivated @samuelcolvin to move to mkdocs, but I personally think it now looks much nicer than the old docs page, and hopefully feels a little more modern for prospective new users.

And at least one person has commented positively (from the fastapi gitter):

> btw, just realized pydantic reached 1.0 and had a very nice facelift for their doc!

------

@haizaar Do you have any issues with new docs format?

**Comment by haizaar** [comment]
@dmontagu not at all. They are aesthetically pleasing. Though there are nice sphinx templates our there as well. And of course you can create hierarchical docs with sphinx as well. I wonder if material design was the main thing. Or may be Samuel just wanted get some experience with mkdocs :)

Again, I have absolutely no problems/issues with the docs. Just being curious.

**Comment by samuelcolvin** [comment]
## 1. Markdown vs. reStructuredText

Many developers find markdown much more intuitive to use than rst. I know rst has many powerful features but it's also extraordinarily convoluted to do simple things e.g. links are

    `name <https://www.example.com/>`_

I find this amazingly complex. Even after having written thousands of lines of it, I still have to search for an example of things that are trivial in md.

I hope people will be more willing to help with the docs now they're written in markdown.

## 2. mkdocs vs. sphinx

mkdocs definitely isn't perfect, there are some design decisions I strongly disagree with. However overall it's a lot more use friendly than sphinx: `mkdocs serve` alone makes a significant difference to the feedback loop when writing docs. I think there's also a richer ecosystem of themes available for mkdocs.

---

> I wonder if material design was the main thing.

Not really, though I do think it's nice.

> Or may be Samuel just wanted get some experience with mkdocs :)

If only, we used a mkdocs site for the help site of the company I run for many years. mkdocs is an old frenemy.

---

In short:
1. I enjoy writing the docs. now more than before.
1. I enjoy reading the docs. now more than before.

**Comment by haizaar** [comment]
Thanks for sharing! Appreciate that! I saw mkdocs becoming quite popular in Python world and was very curious to hear experience from someone who designed to switch. Thanks again!
